src/VecSim/algorithms/hnsw/hnsw.h

```diff
@@ -1381,7 +1402,11 @@ void HNSWIndex<DataType, DistType>::appendVector(const void *vector_data, const
 
     // Access and update the index global data structures with the new vector.
     std::unique_lock<std::mutex> index_data_lock(index_data_guard_);
```

Comment:
In the tiered index scenario, we lock the `index_data_lock` here for the second time (the first time is when we decide whether we should resize the index or not). Consider moving all the logic here to the tiered index and avoid the two times locking...